src/structs.py

```diff
@@ -1333,6 +1333,144 @@ class PathToStateResponse(Response):
+        ground_nsrt = self.nsrt.ground(nsrt_objects)
+        return _GroundLDLRule(self, list(objects), state_pre, goal_pre,
+                              ground_nsrt)
+
+    @cached_property
+    def _str(self) -> str:
+        nsrt_param_str = ", ".join([str(v) for v in self.nsrt.parameters])
+        return f"""LDLRule-{self.name}:
+    Parameters: {self.parameters}
+    State Pre: {sorted(self.state_preconditions, key=str)}
+    Goal Pre: {sorted(self.goal_preconditions, key=str)}
+    NSRT: {self.nsrt.name}({nsrt_param_str})"""
+
+    @cached_property
+    def _hash(self) -> int:
+        return hash(str(self))
+
+    def __str__(self) -> str:
+        return self._str
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __eq__(self, other: object) -> bool:
+        assert isinstance(other, LDLRule)
+        return str(self) == str(other)
+
+    def __lt__(self, other: object) -> bool:
+        assert isinstance(other, LDLRule)
+        return str(self) < str(other)
+
+    def __gt__(self, other: object) -> bool:
+        assert isinstance(other, LDLRule)
+        return str(self) > str(other)
+
+
+@dataclass(frozen=True, repr=False, eq=False)
+class _GroundLDLRule:
+    """A ground LDL rule is an LDLRule + objects.
+
+    Should not be instantiated externally.
+    """
+    parent: LDLRule
+    objects: Sequence[Object]
+    state_preconditions: Set[GroundAtom]
+    goal_preconditions: Set[GroundAtom]
+    ground_nsrt: _GroundNSRT
+
+    @cached_property
+    def _str(self) -> str:
+        nsrt_obj_str = ", ".join([str(o) for o in self.ground_nsrt.objects])
+        return f"""GroundLDLRule-{self.name}:
+    Parameters: {self.objects}
+    State Pre: {sorted(self.state_preconditions, key=str)}
+    Goal Pre: {sorted(self.goal_preconditions, key=str)}
+    NSRT: {self.ground_nsrt.name}({nsrt_obj_str})"""
+
+    @cached_property
+    def _hash(self) -> int:
+        return hash(str(self))
+
+    @property
+    def name(self) -> str:
+        """Name of this ground LRL rule."""
+        return self.parent.name
+
+    def __str__(self) -> str:
+        return self._str
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __eq__(self, other: object) -> bool:
+        assert isinstance(other, _GroundLDLRule)
+        return str(self) == str(other)
+
+    def __lt__(self, other: object) -> bool:
+        assert isinstance(other, _GroundLDLRule)
+        return str(self) < str(other)
+
+    def __gt__(self, other: object) -> bool:
+        assert isinstance(other, _GroundLDLRule)
+        return str(self) > str(other)
+
+
+@dataclass(frozen=True)
+class LiftedDecisionList:
+    """A goal-conditioned policy from abstract states to ground NSRTs
+    implemented with a lifted decision list.
+
+    Given an abstract state and goal, the rules are grounded in order. The
```

Comment:
get rid of this para and the next? it's redundant with `utils.query_ldl` and it's also confusing bc this isn't a function